src/watertap_contrib/reflo/costing/watertap_reflo_costing_package.py

```diff
@@ -73,10 +118,145 @@ def build_global_params(self):
         self.base_currency = pyo.units.USD_2021
 
         # Fix the parameters
-        self.fix_all_vars()
+        # self.fix_all_vars()
         self.plant_lifetime.fix(20)
         self.utilization_factor.fix(1)
-        self.electricity_cost.fix(0.0718)
+        self.electricity_cost.fix(0.0)
+
+        self.electricity_cost_buy = pyo.Param(
+            mutable=True,
+            initialize=0.07,
+            doc="Electricity cost to buy",
+            units=pyo.units.USD_2018 / pyo.units.kWh,
+        )
+
+        self.electricity_cost_sell = pyo.Param(
+            mutable=True,
+            initialize=0.05,
+            doc="Electricity cost to sell",
+            units=pyo.units.USD_2018 / pyo.units.kWh,
+        )
+
+        self.heat_cost_buy = pyo.Param(
+            mutable=True,
+            initialize=0.07,
+            doc="Heat cost to buy",
+            units=pyo.units.USD_2018 / pyo.units.kWh,
+        )
+
+        self.heat_cost_sell = pyo.Param(
+            mutable=True,
+            initialize=0.05,
+            doc="Heat cost to sell",
+            units=pyo.units.USD_2018 / pyo.units.kWh,
+        )
+
+        # Heat balance of the system for sales and purchases of heat
+        treat_cost = self._get_treatment_cost_block()
+        en_cost = self._get_energy_cost_block()
+
+        if all(hasattr(b, "aggregate_flow_heat") for b in [treat_cost, en_cost]):
+            self.frac_heat_from_grid = pyo.Var(
+                initialize=0,
+                domain=pyo.NonNegativeReals,
+                doc="Fraction of heat from grid",
+                units=pyo.units.dimensionless,
+            )
+
+        self.frac_elec_from_grid = pyo.Var(
+            initialize=0,
+            domain=pyo.NonNegativeReals,
+            doc="Fraction of heat from grid",
+            units=pyo.units.dimensionless,
+        )
+
+        self.aggregate_flow_electricity_purchased = pyo.Var(
+            initialize=0,
+            domain=pyo.NonNegativeReals,
+            doc="Aggregated electricity consumed",
+            units=pyo.units.kW,
+        )
+
+        self.aggregate_flow_electricity_sold = pyo.Var(
+            initialize=0,
+            domain=pyo.NonNegativeReals,
+            doc="Aggregated electricity produced",
+            units=pyo.units.kW,
+        )
+
+        self.aggregate_flow_heat_purchased = pyo.Var(
+            initialize=0,
+            domain=pyo.NonNegativeReals,
+            doc="Aggregated heat consumed",
+            units=pyo.units.kW,
+        )
+
+        self.aggregate_flow_heat_sold = pyo.Var(
+            initialize=0,
+            domain=pyo.NonNegativeReals,
+            doc="Aggregated heat produced",
+            units=pyo.units.kW,
+        )
+
+        # energy producer's electricity flow is negative
+        self.aggregate_electricity_balance = pyo.Constraint(
```

Comment:
I added a check so that a ValueError is raised if there is not energy block. Essentially REFLOSystemCosting can only be used if there are both TreatmentCosting and EnergyCosting packages available.